lisp/lisplib/libhelp/libhelp.go

```diff
@@ -0,0 +1,199 @@
+// Copyright © 2021 The ELPS authors
+
+package libhelp
+
+import (
+	"fmt"
+	"io"
+	"strings"
+
+	"github.com/luthersystems/elps/lisp"
+	"github.com/luthersystems/elps/lisp/lisplib/internal/libutil"
+	"github.com/muesli/reflow/dedent"
+	"github.com/muesli/reflow/indent"
+	"github.com/muesli/reflow/wordwrap"
+)
+
+// DeafultPackageName is the package name used by LoadPackage.
+const DefaultPackageName = "help"
+
+// LoadPackage adds the math package to env
+func LoadPackage(env *lisp.LEnv) *lisp.LVal {
+	name := lisp.Symbol(DefaultPackageName)
+	e := env.DefinePackage(name)
+	if !e.IsNil() {
+		return e
+	}
+	e = env.InPackage(name)
+	if !e.IsNil() {
+		return e
+	}
+	for _, op := range ops {
+		env.AddSpecialOps(true, op)
+	}
+	return lisp.Nil()
+}
+
+var ops = []*libutil.Builtin{
+	libutil.FunctionDoc("help", lisp.Formals("var-name"), opHelp,
+		`
+		Prints documentation for the given variable name.  Functions have their
+		signature and any docstring rendered.  Other variables have their types
+		and current values printed.
+		`),
+	libutil.FunctionDoc("help-package", lisp.Formals("pkg-name"), opHelpPackage,
+		`
+		Prints documentation for exported symbols in the specified package.
+		`),
+	libutil.FunctionDoc("help-package-symbols", lisp.Formals("pkg-name", lisp.OptArgSymbol, "all"), opPackageSymbols,
+		`
+		Prints symbols defined in the specified package.  If a second argument
+		is given which evaluates as true then unexported symbols in the package
+		will also be printed.
+		`),
+}
+
+func opHelp(env *lisp.LEnv, args *lisp.LVal) *lisp.LVal {
+	name := args.Cells[0]
+	if name.Type != lisp.LSymbol {
+		return env.Errorf("argument is not a symbol: %v", lisp.GetType(name))
+	}
+	err := RenderVar(env.Runtime.Stderr, env, name.Str)
+	if err != nil {
+		return env.Error(err)
+	}
+	return lisp.Nil()
+}
+
+func opHelpPackage(env *lisp.LEnv, args *lisp.LVal) *lisp.LVal {
+	name := args.Cells[0]
+	if name.Type != lisp.LSymbol {
+		return env.Errorf("argument is not a symbol: %v", lisp.GetType(name))
+	}
+	err := RenderPkgExported(env.Runtime.Stderr, env, name.Str)
+	if err != nil {
+		return env.Error(err)
+	}
+	return lisp.Nil()
+}
+
+func opPackageSymbols(env *lisp.LEnv, args *lisp.LVal) *lisp.LVal {
+	name := args.Cells[0]
+	if name.Type != lisp.LSymbol {
+		return env.Errorf("argument is not a symbol: %v", lisp.GetType(name))
+	}
+	printAll := env.Eval(args.Cells[1])
+	if printAll.Type == lisp.LError {
+		return printAll
+	}
+	pkg := env.Runtime.Registry.Packages[name.Str]
+	if pkg == nil {
+		return env.Errorf("no package: %q", name)
+	}
+	if lisp.True(printAll) {
+		for sym := range pkg.Symbols {
+			_, err := fmt.Fprintln(env.Runtime.Stderr, sym)
+			if err != nil {
+				return env.Error(err)
+			}
+		}
+	} else {
+		exports := pkg.Externals
+		for _, exsym := range exports {
```

Comment:
Looks like this should probably be ordered somehow if possible before printing to match the output without `printAll` being true.

Reply:
oh. I think I fixed a different line number with this. oops.

I actually made the printAll version deterministic because it was ranging over a map. This one should be deterministic, it's just not sorted.

I'm a little torn on whether or not to respect the definition-order of exported functions. I think this is the correct call because you might group a set of functions together logically even if their names aren't lexically adjacent. At least that's my reasoning at the moment.

Reply:
Ah.  I guess I was comparing to `go doc` which appears to sort things alphabetically.